### DISCUSSION TITLE
How to speed up installing dependencies?

### DISCUSSION THREAD

**Discussion by HPDell** [microsoft/vcpkg#36753] [Q&A]
I'm trying to build [QGIS](https://github.com/qgis/QGIS) with Vcpkg. I wrote the following manifest 

```json
{
  "name": "qgis",
  "version": "3.28.8",
  "dependencies": [
    {
      "name" : "vcpkg-cmake",
      "host" : true
    },
    {
      "name" : "vcpkg-cmake-config",
      "host" : true
    },
    {
      "name" : "vcpkg-tool-ninja",
      "host" : true
    },
    "exiv2",
    "gdal",
    "geos",
    "gsl",
    "libspatialindex",
    "libzip",
    "ocilib",
    "proj",
    "protobuf",
    "qca",
    "qscintilla",
    "qt",
    "qtkeychain-qt6",
    "qwt",
    "libspatialite",
    "sqlite3"
  ]
}
```

However, it took me about 2 days so far to install these dependencies. What's worse, my laptop (windows 11) crashed several times when installing some packages. I'm still struggling to build these packages.

So how can I make it faster to install Vcpkg packages?

**Comment by Neumann-A** [comment]
`"qt",` change that to only use the qt dependencies/modules you actually require.

**Reply by HPDell** [reply]
Thanks! I figured this out a few days ago and changed it to `qt-base`. That helped me a lot. Because installing the `qtwebengine` is really slow. But I got some other problems like QGIS cannot find the `Qt Extra` module (of course because this module has been removed since Qt 6.0.0). Nevertheless, these issues are related to QGIS it self, not Vcpkg. But anyway, I think your suggestion helps me a lot in installing dependencies. Thank you again.

**Comment by SchaichAlonso** [comment]
Hi and thanks for working on the qgis port.


Since you use manifest mode, you can set the cmake variables `VCPKG_INSTALL_OPTIONS` to `--debug` and `VCPKG_VERBOSE` to `ON` in the cmake invokation's command line, or in a [CMakePresets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html) file, as vcpkg is invoked in the toolchain, i.e. before the `CMakeList.txt` is parsed.

Setting those variables will cause vcpkg to emit a lot of diagnostics about the packages it builds, and "compatibility data" used to evaluate the binarycache identifier, or [abi-hash](https://learn.microsoft.com/en-us/vcpkg/users/binarycaching#abi-hash), for a package.

With verbose debug enabled, `abientries` blobs will be emited for each package.
E.g. for openssl a blob like
```
[DEBUG] <abientries for openssl:x64-freebsd>
[DEBUG]   asm-comments.patch|c3b4d97d7bf230d934afadfb61e2b88c2d661ecbf8738350bc8da5a40471a9c2
[DEBUG]   cmake|3.28.2
[DEBUG]   declspec-align.patch|6462c7ab198ee97fa5148cd52bde0687cf0d1c59ccb7647801223cca551119e8
[DEBUG]   execute_process|66a937b9c074422643135c319d1abadaa45484a664f1b160d4c163efb444a446
[DEBUG]   features|core;weak-ssl-ciphers
[DEBUG]   install-pc-files.cmake|001f4b55b3da7b4faa0910b906f9c587055abaa99edac9cfb3669a06b5387d44
[DEBUG]   openssl.pc.in|add49223f08228831ced1bc35fc2c56a4ab0fe9741e1fb2bdc86c6d3da4b326a
[DEBUG]   portfile.cmake|38c96858a4eec808c18faa920b1adbf7cfd6e111835328006ff0a1b6b9bb621e
[DEBUG]   ports.cmake|acdf95a924917ff29a9e6bfbee2ecbdb786501f36a847bdff366e41544465592
[DEBUG]   post_build_checks|2
[DEBUG]   qt-msvc.patch|aa76ad678825806fcc4318c7876d2e520d9e5d2a422ce3dfb8f2697d80e20202
[DEBUG]   script-prefix.patch|a18b53acdea8180aa24c4abbe85c4269e4d092c6eba4f946d5feafa7ca2a163d
[DEBUG]   triplet|x64-freebsd
[DEBUG]   triplet_abi|bb8e93cb86245fb38c71ec3f196cc5bff93f4e62dadf91dd7c66237c6d39e2d3-e3caae71f16bbc04ecb1157b7d7ba3d3a07416ea8040a5900c464be7acbf661d-6caeddfbe7d57af3d320b8ec9eabbb3cb53bbfc9
[DEBUG]   unix/android-cc.patch|a3cde12ca3c7c7894e73d300e0ccf728de28b87afde13776e7d7267979c3ead2
[DEBUG]   unix/configure|96ef16490c8a7c240a0c96de7e641c93843d0b78400b4de7dc2c7d2e500f3480
[DEBUG]   unix/disable-quicserver.patch|327cf1f1bff6db9bccaf34381668c1dc556cb08bee9123c0851c6e35dad274a2
[DEBUG]   unix/move-openssldir.patch|9e62ad861f8421e5b732bf488b7199131ac17a57f295f50c697d2591f5845e0e
[DEBUG]   unix/no-empty-dirs.patch|a03d46642c903c9eb09184eabc553ce1467e000c28a9e7d7e92d995970004a62
[DEBUG]   unix/no-static-libs-for-shared.patch|dc51d597ad7477fc56c2a394c84493c87a4de2b84898497dda2754dd0fd8cb53
[DEBUG]   unix/portfile.cmake|3ca966fa7e9380cd5b8a428f2ee070654881daa291815cdcd5a700d0c5d73071
[DEBUG]   unix/remove-deps.cmake|6f9560155faa36f82e249d20b80d28edebd546c55db144b8dabbe0affea6ab14
[DEBUG]   usage|a0bd98264517147f1c61a660a02628e8633b53933c2d35285af49e187081d9d5
[DEBUG]   vcpkg-cmake|bbf1fd8ec9a7011ff0a030883140fb37fdc700f742b9c00854030eef04dc862f
[DEBUG]   vcpkg-cmake-config|8224b21e9cc57a44f91f3021d92440650b37f4f6f03135ac97df9ecb32043390
[DEBUG]   vcpkg-cmake-get-vars|4b7ebb7b7d4fb63666360a44606cf5b96a9f786f2aa2ab2f545d3549a450a06d
[DEBUG]   vcpkg-cmake-wrapper.cmake.in|b29a6ba88f6f26f03a70cf57c31beb6de58ba0306ce8f3028f52fef4b86aec13
[DEBUG]   vcpkg.json|db2382defa39056efbd2a982307b5084f7368bed632dcfcd1dc3427fbb646127
[DEBUG]   vcpkg_acquire_msys|37a90e295fd58dffa4d265c1ce34af0164d1fd844bae8692cd90a6aca1df1ad2
[DEBUG]   vcpkg_add_to_path|5f5ae75cf37b2a58d1a8561ca96496b64cd91ec9a0afab0b976c3e5d59030bfe
[DEBUG]   vcpkg_build_nmake|918422e54a328b9dce1be8c06916d612de9cc24a18c65cc47f26d78e8c487440
[DEBUG]   vcpkg_check_linkage|2236edfea13af20524c17d73214e0e5111bf1c13a2fc46c7f2912a46abd0c1a8
[DEBUG]   vcpkg_configure_make|4811b8a00c6c6371497cd40d02633a89e038b46c24685d4bbf41c3e0bfa0dcfd
[DEBUG]   vcpkg_copy_pdbs|d57e4f196c82dc562a9968c6155073094513c31e2de475694143d3aa47954b1c
[DEBUG]   vcpkg_copy_tools|3d45ff761bddbabe8923b52330168dc3abd295fa469d3f2e47cb14dce85332d5
[DEBUG]   vcpkg_find_acquire_program|96d9ee675798a3bddb54340d404b814be45a35c624b2eab5465cd45eaf18b6d3
[DEBUG]   vcpkg_find_acquire_program(CLANG)|bc9f8433d5ce5a51e45a79c7469cdd9fcad154488b33ff6353af3b29400af50d
[DEBUG]   vcpkg_find_acquire_program(NASM)|71e81fc7dde7b2e8a0c1dc576d44b5898e64f7d6c92d49ce53bd37a4e6929207
[DEBUG]   vcpkg_find_acquire_program(PERL)|74fb0163458bbc0133de10acbdc27137de6ac3cf15a39568b61a1d847e48ff68
[DEBUG]   vcpkg_fixup_pkgconfig|588d833ff057d3ca99c14616c7ecfb5948b5e2a9e4fc02517dceb8b803473457
[DEBUG]   vcpkg_from_git|8f27bff0d01c6d15a3e691758df52bfbb0b1b929da45c4ebba02ef76b54b1881
[DEBUG]   vcpkg_from_github|b743742296a114ea1b18ae99672e02f142c4eb2bef7f57d36c038bedbfb0502f
[DEBUG]   vcpkg_host_path_list|1e8382b3a80fc7458293fce1659077dcacda9892eb63d911ae0330fdb1aedc2d
[DEBUG]   vcpkg_install_copyright|ba6c169ab4e59fa05682e530cdeb883767de22c8391f023d4e6844a7ec5dd3d2
[DEBUG]   vcpkg_install_make|8d8145837ce718ef60778b50fea1af272146c9efa42367ec2d7726332ef29ab5
[DEBUG]   vcpkg_list|f5de3ebcbc40a4db90622ade9aca918e2cf404dc0d91342fcde457d730e6fa29
[DEBUG]   windows/install-layout.patch|06bf9d3f9d78a99d6472bd8a63442065a90cf413af5c4e2946348bece1ce50f2
[DEBUG]   windows/install-pdbs.patch|5ad1f60ee8e967a079ba3d1eed3e6f368f39e5860dd33928594fd0dd58a5ba4d
[DEBUG]   windows/portfile.cmake|812a64310844bf52e31c3d4465ccf7c6045d707b9f830f7715dc7f26396a4482
[DEBUG] </abientries>
```
is emited over here.

Most of the hashes printed are dependent on the revision of the vcpkg repository that is being used.
The `cmake|3.28.2` line indicates that cmake version 3.23.2 is used.
The `features` line corresponds to the `features` keyword of your dependency declaration in `vcpkg.json`, or features that are implied or defaulting.
The `triplet_abi` line contains a hash indicating the compiler, enforced cflags (those used by vcpkg, not those of your project), and so on.
For higher end ports, a hash for each dependency port will also be emited, corresponsing to the abi-hash of the dependency.
A common issue on windows systems is the powershell verion mismatching between systems.

As long as all abientries match between two builds, the [generated binarycache](https://learn.microsoft.com/en-us/vcpkg/users/binarycaching) can be reused. VCPKG defaults to storing binary caches in your home directory, however, it can also be setup to publish binary caches to repositories in the internet. Nuget is popular for this purpose, because you can publish nuget packages for free on github.

---

You can take a free [github actions](https://docs.github.com/en/actions) runner, and configure it up to upload generated binarycaches on github.
The [official documentation provides an example](https://learn.microsoft.com/en-us/vcpkg/users/binarycaching#quickstart-github) on how to setup a runner to configure vcpkg to publish it's binarycache onto github's nuget repository. 

While the free runner isn't exactly impressive in terms of performance, you can use the time to check the verbose debug diagnostics it emits, compare them to your local machine's ones, and try to setup your local machine to use the same configuration as the free runner has - or of cause, to setup the runner to use the same configuration your local machine has.

Once your abientries and the ones from the runner look identical, generate a [personal access token for github](https://github.com/settings/tokens) that has the `read:packages` permission and invoke the same nuget calls you had your github runner execute, but with your personal access token replacing `${{ secrets.GITHUB_TOKEN }}`, then setup the `VCPKG_BINARY_SOURCES` environment variable, like the runner does, on your machine, and ask your vcpkg to deploy. It _should_ attempt to connect to github and download the nuget packages that were previously built by the github actions runner.

As vcpkg checks for the presence of the nuget packages on github when starting to deploy, rather then rechecking after each package was deployed, your local setup will decide to rebuild packages that the actions runner had not yet published when the local machine invoked vcpkg, which will cause vcpkg to suffer a non-fatal exception when attempting to publish from your local machine (if you didn't grant `write:packages` to the token, it will be revoked the publishing attempt anyways). *If* the package was already published but your local machine nevertheless wants to rebuild it, it either failed to download (access token issue or nuget wasn't configured or you forgot to setup the `VCPKG_BINARY_SOURCES` variable) the package, or the abi-hash is mismatching for the package.

Play a bit with the stuff and you will notice that the github actions cannot build qtbase, because it takes just too much disk space for the free runner's limited resources. If you're so far happy with the free runner's work on the "small fries" packages, you can upgrade the free runner to a payed one, which has more resources and is capable of building Qt and the other stuff. I'd recommend you to not invest into those before you managed to get a limited setup running with the free runner up though.

Alternately, you can also use a virtual private server on some inexpensive public cloud system, and configure it as a self-hosted runner, which might or might not be more economic then the github hosted runners. Github hosted free runners are maintained by github, so they're updated by others. Though there are [changelogs](https://github.com/actions/runner-images/releases) and [lists of installed software](https://github.com/actions/runner-images/blob/main/images/windows/Windows2022-Readme.md) of the public runners, they still update "on their own" causing the abi-hashes to break and having to redeploy. Self-Hosted runners don't update w/o you updating them, which is both a blessing and a curse.

TLDR: Build on the cloud and use the nuget binarycaching.

**Reply by HPDell** [reply]
Thanks for your suggestions. So basically, I have to use the binary cache feature to speed up the installation? However, I doubt that GitHub runners can finish installing and caching dependencies in 6 hours, the maximum running time.
